src/hooks/useDebounce.js

```diff
@@ -0,0 +1,42 @@
+import { useState, useEffect, useCallback } from 'react'
+import debounce from 'lodash.debounce'
+
+/**
+ * Custom hook to debounce a function for a specified amount of time
+ */
+const useDebounce = (fn, debounceDurationMs) => {
+  const [isDebounced, setIsDebounced] = useState(false)
+
+  useEffect(() => {
+    let timer
+    if (isDebounced) {
+      timer = setTimeout(() => {
+        setIsDebounced(false)
+      }, debounceDurationMs)
+    }
+    return () => {
+      clearTimeout(timer)
+    }
+  }, [isDebounced])
+
+  let debouncedFn = useCallback(
+    typeof fn === 'function' && debounceDurationMs > 0
+      ? debounce(
```

Comment:
Is this necessary? The outer `setTimeout` + the button's `isDisabled` state already debounces the _button_. i.e.

```ts
const debouncedFn = useCallback(() => {
  if (isDebounced) return;
  setIsDebounced(true);
  fn()
}, [fn, debounceDurationMs])
```

Reply:
> Have you also looked at wrapping lodash.debounce wrapping onClick? May not need use React state

I see maybe this was your initial implementation, which changed in response to this. I think the answer is you need react state because you want the button's presenrtation to change depending on the debounce state. Also, because the debounced version of the function is encapsulated inside the button (i.e. it returns a wrapped local version), there's no possibility of sharing the debounced version. Technically there is now, but I think that use case is undesireable without a concrete example to ensure it has proper test coverage (i.e. some parent component making a debounced version and sharing with multiple child components).

Reply:
i discovered it was necessary after implementing unit tests. when I did not use `lodash.debounce` and relied on the state, I could invoke the "debounced" function multiple times before the state/flag was updated. I couldn't click it fast enough to reproduce this by hand, but I thought it would be safer ensure the function was debounced by definition